### PR TITLE
Add is_empty() for Path and OsStr

### LINKED ISSUE #30259: `PathBuf` does not have `is_empty()`
It is possible to create an empty `PathBuf`:

    let p = PathBuf::new();

— yet `PathBuf` has no `is_empty()` function (nor, apparently, do `Path` or `OsStr` or `OsString`).

This seems inconsistent.

Of course, there are several ways around this (`p == PathBuf::new()`, `p.to_str().map_or(true, |s| s.is_empty())`, wrapping with `Option`), but having a simple way to check this would be nice.

### PR DESCRIPTION
**Pull request by frewsxcv**
Related to https://github.com/rust-lang/rust/pull/30259

Rebase of https://github.com/rust-lang/rust/pull/30623

### CODE REVIEW

File: src/libstd/ffi/os_str.rs [hunk 1]
```diff
@@ -277,6 +277,12 @@ impl OsStr {
         self.to_bytes().and_then(|b| CString::new(b).ok())
     }
 
+    /// Checks if the string is empty.
+    #[unstable(feature = "os_extras", reason = "recently added", issue = "30259")]
```
Comment (nagisa) [inline src/libstd/ffi/os_str.rs]: Should decide between keeping #30259 as a tracking issue for this unstable feature or closing it. I’d rather repurpose the issue in question to be a tracking issue.

Also, perhaps a more self-explanatory feature name would be better? Something like `path_is_empty`?

**Comment by alexcrichton** [comment]
Thanks @frewsxcv! I'm tagging with `T-libs` to ensure this comes up during triage.

File: src/libstd/lib.rs [hunk 2]
```diff
@@ -243,6 +244,7 @@
 #![feature(on_unimplemented)]
 #![feature(oom)]
 #![feature(optin_builtin_traits)]
+#![feature(os_extras)]
```
Comment (alexcrichton) [inline src/libstd/lib.rs]: Were these necessary? I thought that intra-crate usage of unstable features didn't require these annotations...

**Comment by frewsxcv** [comment]
Comments have been addressed

**Comment by eefriedman** [comment]
See also #29453?

**Comment by alexcrichton** [comment]
The libs team discussed this during triage today (sorry for taking awhile to get around to this @frewsxcv!), and we're definitely ok with this, especially in light of #29453.

@frewsxcv that issue is actually tracking an RFC which broadens the methods added here to others like `len` and `with_capacity`, out of curiosity, would you be up for implementing those as well? If not that's ok and I'll just go ahead and send this to bors!

**Comment by frewsxcv** [comment]
Sure! I'll make those changes

**Comment by alexcrichton** [comment]
Ok, cool, thanks! The full RFC is [here](https://github.com/rust-lang/rfcs/blob/master/text/1307-osstring-methods.md), and feel free to ping me if you have any questions!

**Comment by frewsxcv** [comment]
Opened a new PR that covers `OsString` and `OsStr`: https://github.com/rust-lang/rust/pull/31608

**Comment by alexcrichton** [comment]
Thanks @frewsxcv! I'm gonna close this in favor of that.

**Comment by frewsxcv** [comment]
@alexcrichton Should there be a RFC opened for `Path::is_empty`?

**Comment by frewsxcv** [comment]
Opened a PR there: https://github.com/rust-lang/rust/issues/31611

**Comment by alexcrichton** [comment]
Oh right, forgot about that! An RFC issue sounds good to me, though, thanks!

**Pull request closed without merging**